synse/websocket.go

```diff
@@ -0,0 +1,515 @@
+			ID:    addCounter(),
+			Event: requestRead,
+		},
+		Data: opts,
+	}
+
+	resp := new(scheme.ResponseReading)
+	err := c.makeRequest(req, resp)
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.verifyResponse(req.EventMeta, resp.EventMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	return &resp.Data, nil
+}
+
+// ReadDevice returns data from a specific device.
+// It is the same as Read() where the label matches the device id tag
+// specified in ReadOptions.
+// NOTE - this method is not applicable for websocket client.
+func (c *websocketClient) ReadDevice(id string, opts scheme.ReadOptions) (*[]scheme.Read, error) {
+	return nil, nil
+}
+
+// ReadCache returns stream reading data from the registered plugins.
+func (c *websocketClient) ReadCache(opts scheme.ReadCacheOptions) (*[]scheme.Read, error) {
+	req := scheme.RequestReadCache{
+		EventMeta: scheme.EventMeta{
+			ID:    addCounter(),
+			Event: requestReadCache,
+		},
+		Data: opts,
+	}
+
+	resp := new(scheme.ResponseReading)
+	err := c.makeRequest(req, resp)
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.verifyResponse(req.EventMeta, resp.EventMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	return &resp.Data, nil
+}
+
+// WriteAsync writes data to a device, in an asynchronous manner.
+// NOTE - this method is not applicable for websocket client.
+func (c *websocketClient) WriteAsync(id string, opts []scheme.WriteData) (*[]scheme.Write, error) {
+	return nil, nil
+}
+
+// WriteSync writes data to a device, waiting for the write to complete.
+// FIXME - unlike http client's one, this does not support bulk write. Hence, there
+// should only be one WriteData value in the request opts. If multiple values are
+// specified, the first one would be picked. Same for the returned Transaction,
+// there should be only one value.
+// NOTE - this is related to compatibility issue mentioned in synse.go about the
+// need of uniforming client api.
+func (c *websocketClient) WriteSync(id string, opts []scheme.WriteData) (*[]scheme.Transaction, error) {
+	req := scheme.RequestWrite{
+		EventMeta: scheme.EventMeta{
+			ID:    addCounter(),
+			Event: requestWrite,
+		},
+		Data: opts[0], // use the first one in the slice.
+	}
+
+	resp := new(scheme.ResponseWriteState)
+	err := c.makeRequest(req, resp)
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.verifyResponse(req.EventMeta, resp.EventMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	// Quick fix for the issue mentioned above: return a slice with a single value.
+	out := []scheme.Transaction{}
+	out = append(out, resp.Data)
+	return &out, nil
+}
+
+// Transactions returns the sorted list of all cached transaction IDs.
+// NOTE - this method is not applicable for websocket client.
+func (c *websocketClient) Transactions() (*[]string, error) {
+	return nil, nil
+}
+
+// Transaction returns the state and status of a write transaction.
+func (c *websocketClient) Transaction(id string) (*scheme.Transaction, error) {
```

Comment:
another spec issue.. if we support getting transaction status, we should probably also support async writes, since those are where the transaction statuses are returned..